dali/operators/math/normalize/normalize_gpu.cu

```diff
@@ -0,0 +1,279 @@
+
+  if (!has_scalar_mean_) {
+    se.add<float>(AllocType::GPU, param_volume);
+  } else {
+    if (ShouldCalcStdDev()) {
+      // StdDev kernel requires the mean to have the same shape as the output.
+      // We can save memory by broadcasting the mean only to the size of the largest sample
+      // and repeat the pointer for all samples.
+      se.add<float>(AllocType::GPU, MaxSampleSize(param_shape_));
+    }
+  }
+
+  if (!has_scalar_stddev_) {
+    se.add<float>(AllocType::GPU, param_volume);
+  }
+
+  // setup and get memory requirements from kernels
+
+  auto &norm = GetNormalizeKernel<OutputType, InputType>();
+  // if stddev is calculated internally, it's already inverse
+  bool scale_is_stddev = !ShouldCalcStdDev();
+  auto req = norm.Setup(ctx, data_shape_, make_span(axes_),
+                        has_scalar_mean_, has_scalar_stddev_, scale_is_stddev);
+
+  if (ShouldCalcMean()) {
+    auto &mean = GetMeanKernel<float, InputType>();
+    auto mean_req = mean.Setup(ctx, data_shape_, make_span(axes_), true, batch_norm_);
+    assert(mean_req.output_shapes[0] == param_shape_);
+    MaxInPlace(req.scratch_sizes, mean_req.scratch_sizes);
+  }
+
+  if (ShouldCalcStdDev()) {
+    auto &stddev = GetInvStdDevKernel<float, InputType>();
+    auto stddev_req = stddev.Setup(ctx, data_shape_, make_span(axes_), true, batch_norm_);
+    assert(stddev_req.output_shapes[0] == param_shape_);
+    MaxInPlace(req.scratch_sizes, stddev_req.scratch_sizes);
+  }
+
+  for (size_t i = 0; i < se.sizes.size(); i++) {
+    se.add<uint8_t>(static_cast<AllocType>(i), req.scratch_sizes[i], 64);
+  }
+
+  alloc_.Reserve(se.sizes);
+}
+
+template <typename OutputType, typename InputType>
+void Normalize<GPUBackend>::RunTyped(DeviceWorkspace &ws) {
+  auto &input = ws.InputRef<GPUBackend>(0);
+  TensorListView<StorageGPU, const InputType> in_view = view<const InputType>(input);
+
+  auto &output = ws.OutputRef<GPUBackend>(0);
+  TensorListView<StorageGPU, OutputType> out_view = view<OutputType>(output);
+
+  int nsamples = input.ntensor();
+
+  cudaStream_t stream = ws.stream();
+  PreallocatedScratchpad scratch = alloc_.GetScratchpad();
+  KernelContext ctx;
+  ctx.scratchpad = &scratch;
+  ctx.gpu.stream = stream;
+
+  // Prepare mean and stddev
+
+  float scalar_mean   = has_scalar_mean_   ? spec_.GetArgument<float>("mean")   : 0;
+  float scalar_stddev = has_scalar_stddev_ ? spec_.GetArgument<float>("stddev") : 1;
+
+  OutListGPU<float> mean_gpu, stddev_gpu;
+
+  if (!has_scalar_mean_) {
+    mean_gpu = scratch.AllocTensorList<AllocType::GPU, float>(param_shape_);
+  } else if (ShouldCalcStdDev()) {
+    mean_gpu = BroadcastMean(ctx, scalar_mean);
+  }
+
+  if (!has_scalar_stddev_) {
+    stddev_gpu = scratch.AllocTensorList<AllocType::GPU, float>(param_shape_);
+  }
+
+  // We can't just Clear() the scratchpad to reuse it, because temporary buffers are also
+  // stored there - so let's make a snapshot of current allocation state and restore it
```

Comment:
Why before, and not clean after instead?

Reply:
Because the scratchpad is not empty at this point - it contains temporary tensor lists.

Reply:
Maybe it is not most efficient regarding the amount of copies, but how about adding some scoped state saver so you can nest it?

Reply:
That's a jolly good idea, done.